sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java

```diff
@@ -683,6 +743,18 @@ public String apply(MatchResult.Metadata input) {
         return input.resourceId().toString();
       }
     }
+
+    private static class ExtractFilenameAndLastUpdateFn
+        implements SerializableFunction<MatchResult.Metadata, KV<String, Long>> {
+      @Override
+      public KV<String, Long> apply(MatchResult.Metadata input) throws RuntimeException {
+        long timestamp = input.lastModifiedMillis();
+        if (0L == timestamp) {
+          throw new RuntimeException("Extract file timestamp failed.");
```

Comment:
Lets be more descriptive here. Most likely, this will come up when a user does something unusual, and clear error messages are super helpful.